### PR TITLE
Refactor audio decoding

### PR DESCRIPTION
**Pull request by dennwc**
Previously WAV decoding and PCM/ADPCM/MP3 state was all coupled together. This PR splits is into separate structures, making it possible to add new audio formats or test existing ones. This could also help swap the audio backend later too.

It also adds a test that verifies that our decoder gets at least the same number of frames as `ffmpeg` when decoding Nox data (reference test table included).

Hashes of decoded audio we get is not the same as with `ffmpeg`, unfortunately. But it's probably only off by a few bits here and there. Will need to check later.

Also, our music decoding results is just a few samples shorter (~100 samples; ~4 ms) comparing to `ffmpeg` decoding. It doesn't happen with dialogs, though.

## Required sign-off

- [x] I confirm that my PR does not contain any commercial or protected assets and/or source code.
- [x] I agree in advance that my codes will be licensed automatically under the Apache License or similar BSD/MIT-like
      open source licenses in case if OpenNox Project will adopt such a non-GPL license in the future.

### CODE REVIEW

File: src/legacy/client/audio/ail/audio_test.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+			exp:  dialogHashes,
+		},
+		{
+			name: "music",
+			exp:  musicHashes,
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			path := noxtest.DataPath(t, c.name)
```
Comment (dennwc) [inline src/legacy/client/audio/ail/audio_test.go:48]: @blmarket You might be interested in this trick here.

This function uses the same logic as OpenNox to find your local Nox installation during tests and open data files from there.

We can not include any Nox data, but we can at least test the _hashes_ of the data we decode, for example.

**Comment by blmarket** [comment]
Seems weird. my comment is not showing if not logged in... Trying to quote:



> 1. Not sure we can ask other devs to have their own installations to run tests. i.e. this change creates external dependency which dev may not have right now
> 2. dialog can be localized, so some files are not always available. (e.g. there can be XYXYE.wav and XYXYK.wav etc.)
>
> Would be great if there's a flag to control running test which needs original game data, and disable it by default. (How about we skip the test if env NOX_DATA is not specified?)

**Comment by dennwc** [comment]
Yeah, it runs under the assumption that there's at least one GoG-like installation somewhere. I do have localizations too, but they are in different folders that auto-detection cannot recognize.

> How about we skip the test if env NOX_DATA is not specified?

The whole point of that `noxtest` package was to not set `NOX_DATA` explicitly :sweat_smile: Maybe some special value will do? Like `NOX_DATA=-`.

**Comment by blmarket** [comment]
Let me take it back. I think current behavior is the best.

What I expected to see:
- (+) Test does not fail simply because user does not have nox data
- (+) Test do some checks if data files are available
- (Maybe need some improvements) test let user know if it couldn't find nox data. (Currently user need to use -v option to see the issue)

**Comment by dennwc** [comment]
Unfortunately I don't think there's any good way around `-v`. Without it all test logs (including skipped tests, as you noticed) are suppressed.

We could write to `stderr` directly to get around that, but this usually breaks other Go tools that expect specific output format from `go test`. We could also write to `stdout` from tests, but that's also considered a bad practice.

So we should probably keep the current behavior and probably document that its advised to pass `-v` to tests.

**Comment by blmarket** [comment]
> Unfortunately I don't think there's any good way around `-v`. Without it all test logs (including skipped tests, as you noticed) are suppressed.
> 
> We could write to `stderr` directly to get around that, but this usually breaks other Go tools that expect specific output format from `go test`. We could also write to `stdout` from tests, but that's also considered a bad practice.
> 
> So we should probably keep the current behavior and probably document that its advised to pass `-v` to tests.

Agreed. thanks for sharing your thoughts!

**Pull request merged**